tools/btmidi-server.c

```diff
@@ -0,0 +1,927 @@
+
+	server->io = io_new(pfd.fd);
+	if (!server->io) {
+		fprintf(stderr, "Could not allocate I/O eventloop\n");
+		goto _err_port;
+	}
+
+	io_set_read_handler(server->io, midi_notify_cb, server, NULL);
+
+	/* Init the MIDI parser */
+	err = midi_read_init(&server->midi_in);
+	if (err < 0) {
+		fprintf(stderr, "Could not initialise MIDI input parser\n");
+		goto _err_port;
+	}
+
+	err = midi_write_init(&server->midi_out, bt_att_get_mtu(server->att) - 3);
+	if (err < 0) {
+		fprintf(stderr, "Could not initialise MIDI output parser\n");
+		goto _err_midi;
+	}
+
+	/* Add MIDI Service */
+	bt_string_to_uuid(&uuid, MIDI_UUID);
+	service = gatt_db_add_service(server->db, &uuid, true, 6);
+
+	/* Add MIDI IO Characteristics */
+	bt_string_to_uuid(&uuid, MIDI_IO_UUID);
+	server->midi_io = gatt_db_service_add_characteristic(service, &uuid,
+						BT_ATT_PERM_READ | BT_ATT_PERM_WRITE,
+						BT_GATT_CHRC_PROP_WRITE_WITHOUT_RESP | BT_GATT_CHRC_PROP_READ | BT_GATT_CHRC_PROP_NOTIFY,
+						NULL, midi_write_cb, server);
+	server->midi_io_handle = gatt_db_attribute_get_handle(server->midi_io);
+
+	/* Add MIDI CCC */
+	bt_uuid16_create(&uuid, GATT_CLIENT_CHARAC_CFG_UUID);
+	gatt_db_service_add_descriptor(service, &uuid,
+					BT_ATT_PERM_READ | BT_ATT_PERM_WRITE,
+					midi_ccc_read_cb, midi_ccc_write_cb, server);
+
+	/* Activate Service */
+	gatt_db_service_set_active(service, true);
+
+	return 0;
+
+_err_midi:
+	midi_read_free(&server->midi_in);
+
+_err_port:
+	snd_seq_delete_simple_port(server->seq_handle, server->seq_port_id);
+
+_err_handle:
+	snd_seq_close(server->seq_handle);
+	server->seq_handle = NULL;
+
+	return -1;
+}
+
+static void add_adv_callback(uint8_t status, uint16_t length,
+							const void *param, void *user_data)
+{
+// 	struct btd_adv_client *client = user_data;
+	const struct mgmt_rp_add_advertising *rp = param;
+
+	if (status)
+		goto done;
+
+	if (!param || length < sizeof(*rp)) {
+		status = MGMT_STATUS_FAILED;
+		goto done;
+	}
+done:
+	printf("Blz\n");
+}
+
+static int advertise(int dev_id)
+{
+	struct mgmt *mgmt_if = NULL;
+	struct bt_ad *data;
+	uint8_t *adv_data;
+	size_t adv_data_len;
+	struct mgmt_cp_add_advertising *cp;
+	uint8_t param_len;
+	bt_uuid_t uuid;
+	
+	bt_string_to_uuid(&uuid, MIDI_UUID);
+
+	mgmt_if = mgmt_new_default();
+	if (!mgmt_if) {
+		fprintf(stderr, "Failed to access management interface\n");
+		return -1;
+	}
+
+	data = bt_ad_new();
+	if (!data) {
+		fprintf(stderr, "Error creating adverting data\n");
+		return -1;
+	}
+
+	if (!bt_ad_add_service_uuid(data, &uuid)) {
+		fprintf(stderr, "Error adding service UUID\n");
+		return -1;
+	}
+
+	adv_data = bt_ad_generate(data, &adv_data_len);
+	if (!adv_data) {
+		fprintf(stderr, "Error generating advertising data\n");
+		return -1;
+	}
+
+	param_len = sizeof(struct mgmt_cp_add_advertising) + adv_data_len;
+	cp = malloc0(param_len);
+	if (!cp) {
+		fprintf(stderr, "Error allocating memory to advertising params\n");
+		return -1;
```

Comment:
I pushed in my repo yesterday night a partial implementation. I didn't know that it would be tracked here even without a new pull request. It is not finished yet. The advertising itself is not OK.